core/src/main/java/org/apache/druid/java/util/common/FileUtils.java

```diff
@@ -207,9 +215,20 @@ private static void writeAtomically(final File file, final File tmpDir, OutputSt
           StandardCopyOption.ATOMIC_MOVE,
           StandardCopyOption.REPLACE_EXISTING
       );
+
+      // fsync the directory entry to ensure the new file will be visible after a crash.
+      try (final FileChannel directory = FileChannel.open(file.getParentFile().toPath(), StandardOpenOption.READ)) {
+        directory.force(true);
+      }
+
+      return retVal;
     }
     finally {
-      tmpFile.delete();
+      if (tmpFile.exists()) {
+        if (!tmpFile.delete()) {
```

Comment:
`Files.delete()` will allow to see the reason of the failure (IO error or file not found, or permissions, etc.)